internal/abci/abci.go

```diff
@@ -815,11 +849,12 @@ func (a *AbciApp) ProcessProposal(ctx context.Context, req *abciTypes.RequestPro
 		zap.Int64("height", req.Height),
 		zap.Int("txs", len(req.Txs)))
 
-	addr := proposerAddrToString(req.ProposerAddress)
-	proposerPubKey, err := a.getValidatorPubKeyByAddr(ctx, addr, &req.Height)
-	if err != nil {
-		a.log.Warn("received block proposal from unknown validator", zap.String("addr", addr))
-		return &abciTypes.ResponseProcessProposal{Status: abciTypes.ResponseProcessProposal_REJECT}, nil
+	proposerPubKey, ok := a.validatorAddressToPubKey[proposerAddrToString(req.ProposerAddress)]
+	if !ok {
+		// there is an edge case where cometbft will allow a node to PrepareProposal
+		// even if it is not a validator, if it was a validator in the most recent block.
```

Comment:
Regardless, while this may be a workaround for not tracking the delayed updates to the canonical validator set, if cometbft calls `ProcessProposal` with a given `ProposerAddress`, that means they are a validator.  CometBFT processes all updates given to it and from that manages the validator set (and all the precommit voting involves its canonical validator set), which includes `ProposerAddress`.

All that to say that we should soften the error since with the current solution it does not indicate a malfunction of the application. It is expected and working as intended, even if it's a workaround.

Reply:
> Are these comments alluding to the h vs h+2 discrepancy, or some other edge where a validator updated to power 0 at height h can actually be asked to prepare a proposal for h+2 if it was (acceptably) the proposer at h+1? So this error reflects an unfixed big in some sense.

Yes